code/datums/components/battery_power.dm

```diff
@@ -0,0 +1,51 @@
+/// Generalized framework allowing something to draw power from an inserted power cell.
+/// Contains attack_hand() and attackby() signals allowing cells to be added and removed by hand.
+/// This base type supports only standard cells; device cells won't fit.
+/datum/component/battery_power
+	var/obj/item/cell/cell
+
+/datum/component/battery_power/Initialize()
+	if (istype(parent, /obj/machinery))
+		RegisterSignal(parent, COMSIG_ATOM_ATTACK_HAND, .proc/on_attack_hand)
+		RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/on_attack_by)
+	else
+		return COMPONENT_INCOMPATIBLE
+
+/// Checks if a provided item can be used as a power cell for this component.
+/// By default, the component checks for standard-sized cells.
+/datum/component/battery_power/proc/is_valid_cell(obj/item/I)
+	return istype(I, /obj/item/cell) && !istype(I, /obj/item/cell/device)
```

Comment:
Oh that's right, this was the awful solution for device cells. God, but I wish this were cleaner. Out of scope, but maybe if we make the current cells a separate subtype of an abstract parent class that has to core cell logic, and machinery is just allowed to use the machinery subtype and items are just allowed to use the device subtype